RFS/src/main/java/com/rfs/common/S3SnapshotCreator.java

```diff
@@ -0,0 +1,152 @@
+
+    private final ConnectionDetails connectionDetails;
+    private final String snapshotName;
+    private final String s3Uri;
+    private final String s3Region;
+
+    public S3SnapshotCreator(String snapshotName, ConnectionDetails connectionDetails, String s3Uri, String s3Region) {
+        this.snapshotName = snapshotName;
+        this.connectionDetails = connectionDetails;
+        this.s3Uri = s3Uri;
+        this.s3Region = s3Region;
+    }
+
+    public String getSnapshotName() {
+        return snapshotName;
+    }
+
+    /*
+     * Extracts the bucket name from the S3 URI
+     * s3://my-bucket-name/my-folder/my-nested-folder => my-bucket-name
+     */
+    public String getBucketName() {
+        return s3Uri.split("/")[2];
+    }
+
+    /*
+     * Extracts the base path from the S3 URI; could be nested arbitrarily deep
+     * s3://my-bucket-name/my-folder/my-nested-folder => my-folder/my-nested-folder
+     */
+    public String getBasePath() {
+        int thirdSlashIndex = s3Uri.indexOf('/', 5);
+        if (thirdSlashIndex == -1) {
+            // Nothing after the bucket name
+            return null;
+        }
+
+        // Extract everything after the third "/", excluding any final "/"
+        String rawBasePath = s3Uri.substring(thirdSlashIndex + 1);
+        if (rawBasePath.endsWith("/")) {
+            return rawBasePath.substring(0, rawBasePath.length() - 1);
+        } else {
+            return rawBasePath;
+        }
+    }
+
+    public void registerRepo() throws Exception {
+        // Assemble the REST path
+        String targetName = "_snapshot/s3_repo";
+
+        // Assemble the request body
+        ObjectNode settings = mapper.createObjectNode();
+        settings.put("bucket", getBucketName());
+        settings.put("region", s3Region);
+        settings.put("base_path", getBasePath());
+
+        ObjectNode body = mapper.createObjectNode();
+        body.put("type", "s3");
+        body.set("settings", settings);
+
+        // Register the repo; it's fine if it already exists
+        RestClient client = new RestClient(connectionDetails);
+        String bodyString = body.toString();
+        RestClient.Response response = client.put(targetName, bodyString, false);
+        if (response.code == HttpURLConnection.HTTP_OK) {
+            logger.info("S3 Repo registration successful");
+        } else {
+            logger.error("S3 Repo registration failed");
+            throw new RepoRegistrationFailed("s3_repo");
+        }
+    }
+
+    public void createSnapshot() throws Exception {
+        // Assemble the REST path
+        String targetName = "_snapshot/s3_repo/" + getSnapshotName();
+
+        // Assemble the request body
+        ObjectNode body = mapper.createObjectNode();
+        body.put("indices", "_all");
+        body.put("ignore_unavailable", true);
+        body.put("include_global_state", true);
+
+        // Register the repo; it's fine if it already exists
```

Comment:
This comment seems like it should be removed

Reply:
I thought it provided value, but I guess it wasn't clear enough.  Will re-write.

Reply:
Done

Reply:
It seemed like it was saying this method would register the repo, but was pretty sure that happened in another method

Reply:
Oh, whoops, sorry - you're completely correct.  I'll change that locally and it will flow out the next time I commit some code.